boot/flags.go

```diff
@@ -0,0 +1,282 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package boot
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/snapcore/snapd/bootloader"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/strutil"
+)
+
+var (
+	errNotUC20 = fmt.Errorf("cannot get boot flags on non-UC20 device")
+
+	understoodBootFlags = []string{
+		// the factory boot flag is set to indicate that this is a
+		// boot inside a factory environment
+		"factory",
+	}
+)
+
+type unknownFlagErr string
+
+func (e unknownFlagErr) Error() string {
+	return string(e)
+}
+
+func IsUnknownBootFlagError(e error) bool {
+	_, ok := e.(unknownFlagErr)
+	return ok
+}
+
+// splitBootFlagString splits the given comma delimited list of boot flags, removing
+// empty strings.
+// Note that this explicitly does not filter out unsupported boot flags in the
+// off chance that an old version of the initramfs is reading new boot flags
+// written by a new version of snapd in userspace on a previous boot.
+func splitBootFlagString(s string) []string {
+	flags := []string{}
+	for _, flag := range strings.Split(s, ",") {
+		if flag != "" {
+			flags = append(flags, flag)
+		}
+	}
+
+	return flags
+}
+
+func checkBootFlagList(flags []string, allowList []string) ([]string, error) {
+	allowedFlags := make([]string, 0, len(flags))
+	disallowedFlags := make([]string, 0, len(flags))
+	if len(allowList) != 0 {
+		// then we need to enforce the allow list
+		for _, flag := range flags {
+			if strutil.ListContains(allowList, flag) {
+				allowedFlags = append(allowedFlags, flag)
+			} else {
+				if flag == "" {
+					// this is to make it more obvious
+					disallowedFlags = append(disallowedFlags, "\"\"")
```

Comment:
```suggestion
					disallowedFlags = append(disallowedFlags, `""`)
```

Reply:
I did this here since I needed a master merge to unbreak arch